cmd/acb/commands/exec/exec.go

```diff
@@ -194,6 +196,32 @@ var Command = cli.Command{
 			}
 		}
 
+		type Wrapper struct {
+			Version string `yaml:"version,omitempty"`
+		}
+
+		wrap := &Wrapper{}
+		if err := yaml.Unmarshal(template.GetData(), wrap); err != nil {
+			return err
+		}
+
+		shouldIncludeAlias := wrap == nil || wrap.Version >= "v.1.1.0"
+		var task *graph.Task
+		if shouldIncludeAlias {
+			// Generate the base task file without resolving environment variables.
+			task, err := graph.NewTaskFromBytes(template.GetData())
+			if err != nil {
+				return err
+			}
+
+			// Remarshall functional task file to resolve templating
+			fromTask, err := yaml.Marshal(task)
+			if err != nil {
+				return err
+			}
+			template.Data = fromTask
+		}
+
 		rendered, err := templating.LoadAndRenderSteps(ctx, template, renderOpts)
```

Comment:
This `rendered` string will have rendered things like: 
{{.Run.Registry}}

Your code never takes this path (i.e. uses rendered to to UnmarshalTaskFromString.

That is a bug. Shouldnt it be in this sequence:

1. Load up all the "Task aliases" in memory (in the `aliasMap`.)
For instance, 
> commit: "{{.Run.Commit}}"
repository: "{{.Run.Repository}}"
branch: "{{.Run.Branch}}"
gitTag: "{{.Run.GitTag}}"
triggeredBy: "{{.Run.TriggeredBy}}"
registry: "{{.Run.Registry}}"
registryName: "{{.Run.RegistryName}}"
date: "{{.Run.Date}}"
sharedVolume: "{{.Run.SharedVolume}}"
os: "{{.Run.OS}}"
architecture: "{{.Run.Architecture}}"
id: "{{.Run.ID}}"
osArchitecture: ""
//  Usable without $
purge: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e purge"
tagList: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e tag list"
manifestList: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e manifest list"
manifestDelete: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e manifest delete"
tagDelete: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e tag delete"
// Usable without directive
acr: "mcr.microsoft.com/acr/acr-cli:linux-7ccdc3e"
az: "mcr.microsoft.com/azure-cli"
bash: "bash" # Note for this one bash is in fact the alias and the source at dockerhub
buildkit:
buildx: ""
curl: "appropriate/curl"
docker: "docker" # Note for this one docker is in fact the alias and the source at dockerhub
dotnet: "mcr.microsoft.com/dotnet/core"
git: ""
go: "golang"
helm: "cmd.azurecr.io/helm:v2.11.0-rc.2"
kubectl: "bitnami/kubectl"
npm: "chekote/node"
wget: "abihf/wget"
yarn: "chekote/node"

And some more...

2) Preprocess the string Task
3) "LoadAndRenderSteps" aka Render any go templating in the string 
4) UnmarshalTaskFromString.

I suggest separating out the "preprocessing" logic from "NewTaskFromBytes". IMO, the Task object should only be made after Rendering is done.

What do you say, @northtyphoon ?

Reply:
Yes but there is an issue with that. The only reason why I had to do this was as a result of the directive-less image names under cmd : In order to process those properly we must have a task already read into a task object so we can determine the location of cmd values. As a result, in order to avoid rendering twice I first fully obtain the task file.